### PR TITLE
fix(web): verify pull request snapshots before hydration

### PR DESCRIPTION
**Pull request by Adamulek123**
## What Changed

Persisted pull request rows now carry a short expiration and are hydrated only after a lightweight server read freshly verifies the signed-in account for every host. The identity read is separate from the repository listing, so safe snapshots can still render while the slower live list is loading.

## Why

Snapshots were scoped only by environment IDs. Switching a host CLI to another account could therefore briefly expose the previous account's pull requests, and an offline host could retain old rows indefinitely. Waiting for the full list would prevent the leak but also defeat early hydration, so account verification now has its own focused RPC.

## UI Changes

No visual design changes. The loading behavior is unchanged for a valid, recent snapshot; expired or differently owned snapshots fall back to the existing loading state.

## Validation

- `vp test run apps/server/src/pullRequest/PullRequestService.test.ts apps/web/src/components/pullRequest/pullRequestList.logic.test.ts packages/contracts/src/pullRequest.test.ts apps/server/src/auth/RpcAuthorization.test.ts` (209 tests passed)
- `vp run --filter @t3tools/contracts typecheck`
- `vp run --filter @t3tools/client-runtime typecheck`
- `vp run --filter t3 typecheck`
- `vp run --filter @t3tools/web typecheck`

## Checklist

- [x] This PR is small and focused
- [x] I explained what changed and why
- [x] No visual UI change requires before/after screenshots
- [x] No animation or interaction change requires a video

Generated with GPT-5.6 in the Codex harness.

---

> [!NOTE]
> **Medium Risk**
> Changes how pull-request rows are authorized and cached across CLI account switches and multi-environment merges; incorrect gating could hide lists or briefly show stale retained rows, though the design explicitly invalidates on identity mismatch.
> 
> **Overview**
> Adds a dedicated **`pullRequests.viewers`** RPC and server **`PullRequestService.viewers`** path that re-reads signed-in host CLI identity with **`fresh: true`** (no viewer TTL cache). When a fresh read disagrees with a cached identity, the server bumps **`listingsEpoch`** so in-flight or cached listings cannot serve the previous account. **`refineUnknownProjectKinds`** now honors **`projectIds`** for that scope.
> 
> The pull-requests route runs **`usePullRequestViewers`** first and **`resolvePullRequestViewerGate`** blocks list/baseline/partition queries—and localStorage hydration—until capable environments verify (or legacy servers without **`pullRequestViewers`** keep live listing only). Snapshots get **`writtenAt`**, a **5-minute** max age, mandatory viewer context on read, and helpers to match, filter, and drop rows when verification fails or accounts change.
> 
> **Behavioral note:** snapshots missing **`writtenAt`** no longer decode; refresh and pagination wait on viewer verification where the capability is advertised.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit aa30e5a33736a52dfcb1beb92f21e08f7ec169a2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Verify pull request viewer identity before snapshot hydration
> - Adds a `viewers` RPC and `PullRequestService.viewers` method to fetch fresh host identities, scoped to environments that advertise the new `pullRequestViewers` capability.
> - `readPullRequestListSnapshot` in [pullRequestList.logic.ts](https://github.com/pingdotgg/t3code/pull/6470/files#diff-f8ca5dc8d0f0bac3b412ba2abaf7fca00497f36286660cfeb38242768c60e478) now requires a viewer context and a `writtenAt` timestamp, rejecting snapshots older than 5 minutes or with mismatched viewer identities.
> - The PR route gates list queries, snapshot hydration, and infinite scroll on verification status, pruning retained rows for environments that fail verification.
> - Behavioral Change: `readPullRequestListSnapshot` and `writePullRequestListSnapshot` signatures changed to require context objects; persisted snapshots missing `writtenAt` will fail to decode.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized aa30e5a.</sup>
> 
>

### CODE REVIEW

File: apps/server/src/pullRequest/PullRequestService.ts [hunk 1]
```diff
@@ -743,6 +754,23 @@ export const make = Effect.gen(function* () {
       { concurrency: REPOSITORY_CONCURRENCY },
     );
 
+  // Snapshot hydration needs identity before the slower repository listing has answered. This
+  // read deliberately bypasses the viewer cache: an account can be switched outside T3, and a
+  // cached identity must never authorize rows persisted for the previous account.
+  const viewers: PullRequestService["Service"]["viewers"] = (input) =>
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/server/src/pullRequest/PullRequestService.ts:779]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 9ea9dd8e. A fresh viewer change or failed fresh identity read now advances the listings epoch before cached rows can be reused, stranding both listCache and staleList entries authorized for the previous account. The regression test proves the next list returns the new viewer; all 90 focused service tests pass.

File: apps/server/src/pullRequest/PullRequestService.ts [hunk 2]
```diff
@@ -735,14 +748,48 @@ export const make = Effect.gen(function* () {
...
             Effect.tap((result) =>
-              Effect.map(Clock.currentTimeMillis, (at) => viewersByHost.set(host, { at, result })),
+              Effect.map(Clock.currentTimeMillis, (at) => {
+                if (
+                  options.fresh === true &&
+                  held !== undefined &&
+                  held.result.viewer !== result.viewer
+                ) {
+                  listingsEpoch = ++epochCounter;
+                }
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/server/src/pullRequest/PullRequestService.ts:758]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 610a6b79. The page now withholds its feed, baseline, and partition list queries until the fresh viewer query has settled, so the server has advanced the listings epoch before any warm-cache key is sampled. The web typecheck, 97 focused list-logic tests, and all 90 focused service tests pass.

File: apps/web/src/state/pullRequests.ts [hunk 3]
```diff
@@ -99,6 +105,27 @@ export interface MergedPullRequestListView {
...
+        : Object.fromEntries(
+            query.values.flatMap(([environmentId, result]) =>
+              Object.entries(result.viewers).map(
+                ([host, viewer]) => [`${environmentId} ${host}`, viewer] as const,
+              ),
+            ),
+          ),
+    [query.values],
+  );
+  return { viewers, isPending: query.isPending };
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/state/pullRequests.ts]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 5a488c4d. The viewers hook now exposes successful environment IDs, errors, and refresh. A total viewer failure renders the unavailable state, and both the header refresh and retry path re-run viewer verification.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 4]
```diff
@@ -623,7 +624,26 @@ function PullRequestsRouteView() {
...
+        },
+      })),
+    [environmentQueries, scopedProjectId, search.host],
+  );
+  const viewerQuery = usePullRequestViewers(viewerTargets);
+  // The fresh viewer read also invalidates server-side listing caches when the CLI account has
+  // changed. Do not race a warm list hit against that read: once this gate opens, every listing
+  // is keyed under the identity the page just verified.
+  const viewerVerified = !viewerQuery.isPending && viewerQuery.viewers !== null;
+  const listQuery = usePullRequestList(viewerVerified ? listTargets : NO_LIST_TARGETS);
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 5a488c4d. The first-load condition now includes the viewer identity round trip, so cold loads keep the skeleton until verified list reads can begin.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 5]
```diff
@@ -623,7 +624,26 @@ function PullRequestsRouteView() {
...
+          ...(search.host ? { host: search.host } : {}),
+        },
+      })),
+    [environmentQueries, scopedProjectId, search.host],
+  );
+  const viewerQuery = usePullRequestViewers(viewerTargets);
+  // The fresh viewer read also invalidates server-side listing caches when the CLI account has
+  // changed. Do not race a warm list hit against that read: once this gate opens, every listing
+  // is keyed under the identity the page just verified.
+  const viewerVerified = !viewerQuery.isPending && viewerQuery.viewers !== null;
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 5a488c4d. Feed, baseline, authored, and reviewing targets are filtered independently to the environments whose viewer RPC succeeded; a partial failure cannot issue a list request for the unverified environment. Web typecheck and all 97 focused list-logic tests pass.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 6]
```diff
@@ -623,7 +624,33 @@ function PullRequestsRouteView() {
...
+  // identity check failed: every listing is keyed under the identity this page just verified.
+  const verifiedEnvironmentIds = useMemo(
+    () => new Set(viewerQuery.environmentIds),
+    [viewerQuery.environmentIds],
+  );
+  const verifiedListTargets = useMemo(
+    () => listTargets.filter(({ environmentId }) => verifiedEnvironmentIds.has(environmentId)),
+    [listTargets, verifiedEnvironmentIds],
+  );
+  const listQuery = usePullRequestList(verifiedListTargets);
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx:682]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 81a65ed3. Successful viewer values are now ignored while the merged viewer query is revalidating, so feed, baseline, and partition list targets all stay empty until the fresh identity round trip has fully settled. Web typecheck and all 97 focused list-logic tests pass.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 7]
```diff
@@ -719,6 +764,7 @@ function PullRequestsRouteView() {
       setInvalidating(false);
     }
     refreshList();
+    viewerQuery.refresh();
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx:796]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 6b937e76. The shared refreshing flag now stays active through viewer revalidation as well as invalidation and list loading, so both refresh controls remain disabled and spinning for the full refresh. Web typecheck and all 97 focused list-logic tests pass.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 8]
```diff
@@ -1336,8 +1386,14 @@ function PullRequestsRouteView() {
         />
       ) : firstLoad ? (
         <PullRequestListGhost rows={7} />
-      ) : listQuery.error && listData === null ? (
-        <PullRequestsUnavailableState error={listQuery.error} onRetry={() => listQuery.refresh()} />
+      ) : listError && listData === null ? (
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx:1423]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 1d9b4b1d. The stale-row banner now uses the combined viewer/list error and its Retry refreshes both queries, so total and partial viewer failures remain visible and recoverable when carried rows are on screen. Web typecheck and all 97 focused list-logic tests pass.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 9]
```diff
@@ -1336,8 +1386,14 @@ function PullRequestsRouteView() {
...
+        <PullRequestsUnavailableState
+          error={listError}
+          onRetry={() => {
+            viewerQuery.refresh();
+            listQuery.refresh();
+          }}
+        />
       ) : carriedToNothing ? (
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx:1431]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 138977db. `carriedToNothing` now treats viewer verification as pending, so filtered carried rows show the loading ghost until identity and list reads settle. The pagination observer is also paused during viewer verification or viewer failure. Web typecheck and all 97 focused list-logic tests pass.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 10]
```diff
@@ -1053,8 +1103,9 @@ function PullRequestsRouteView() {
       !sentinel ||
       entries.length === 0 ||
       listData?.truncated !== true ||
+      viewerQuery.isPending ||
       listQuery.isPending ||
-      listQuery.error !== null ||
+      listError !== null ||
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 7a2f7c7d. Pagination remains paused while viewer verification is pending, but an error from an unverified environment no longer blocks the successfully verified environments; only the active list query's own error stops its sentinel. Web typecheck and all 97 focused list-logic tests pass.

File: apps/web/src/routes/_chat.pull-requests.tsx [hunk 11]
```diff
@@ -623,7 +624,33 @@ function PullRequestsRouteView() {
...
+  // identity check failed: every listing is keyed under the identity this page just verified.
+  const verifiedEnvironmentIds = useMemo(
+    () => new Set(viewerQuery.isPending ? [] : viewerQuery.environmentIds),
+    [viewerQuery.environmentIds, viewerQuery.isPending],
+  );
+  const verifiedListTargets = useMemo(
+    () => listTargets.filter(({ environmentId }) => verifiedEnvironmentIds.has(environmentId)),
+    [listTargets, verifiedEnvironmentIds],
+  );
+  const listQuery = usePullRequestList(verifiedListTargets);
```
[the remark this answers is not in the document]
Comment (Adamulek123) [reply apps/web/src/routes/_chat.pull-requests.tsx:682]: gpt-5.6-sol - responding
=====

Confirmed and fixed in 27934600. The environment descriptor now advertises `pullRequestViewers`; the route calls the viewer RPC only for servers that support it. Older servers retain live listing but cannot hydrate persisted rows, while capable servers still fail closed on identity-check errors. The three focused suites (102 tests) and contracts, web, and server typechecks pass.

**Comment by Adamulek123** [comment]
Added the requested edge-case coverage after rebasing onto the latest `origin/main` (commit `9d6c0932`):

- one failed host alongside one healthy host
- failed viewer verification with an existing cached snapshot
- account switching while a list request is in flight
- old servers that support pull requests but not viewer verification

Local verification:

- `vp test run apps/server/src/pullRequest/PullRequestService.test.ts apps/web/src/components/pullRequest/pullRequestList.logic.test.ts packages/contracts/src/pullRequest.test.ts`: 3 files passed, 214 tests passed
- targeted typechecks for `t3`, `@t3tools/web`, and `@t3tools/contracts`: passed
- targeted lint: no errors; one existing `prefer-set-has` advisory remains in `pullRequestList.logic.ts`
- `git diff --check`: passed

**Comment by Adamulek123** [comment]
Superseded by #9109 and #9110. The work is now split into a viewer-verification foundation and a stacked route integration. The final #6470 state, including its bot-review fixes, was preserved in the replacement branches.
